File: backend/app/services/evaluator.py
from __future__ import annotations

from typing import Any


def classify(score: float, lang: str) -> str:
    labels_ar = [
        (95, "استثنائية ونادرة"), (90, "ممتازة جدًا"), (85, "قوية"), (80, "جيدة"),
        (70, "متوسطة"), (60, "ضعيفة نسبيًا"), (50, "غير مقنعة"), (0, "غير جاهزة مهنيًا"),
    ]
    labels_en = [
        (95, "Exceptional"), (90, "Very strong"), (85, "Strong"), (80, "Good"),
        (70, "Average"), (60, "Relatively weak"), (50, "Unconvincing"), (0, "Not ready"),
    ]
    labels = labels_ar if lang == "ar" else labels_en
    return next(label for threshold, label in labels if score >= threshold)


def evaluate(parsed: dict[str, Any], target_role: str | None, job_description: str | None, lang: str) -> dict[str, Any]:
    sections = parsed["sections"]
    contact = parsed["contact_information"]
    bullets = parsed["bullet_points"]
    numbers = parsed["numbers"]
    links = contact["links"]

    main = 0.0
    internal = 0.0
    issues: list[dict[str, Any]] = []
    strengths: list[dict[str, Any]] = []

    # Main criteria placeholder rubric, intentionally conservative.
    main += 6 if target_role else 4
    main += 5 if len(bullets) >= 4 else 3
    main += min(5, 1 + len(numbers) * 0.6)
    main += 3 if sections.get("experience") or sections.get("projects") else 1
    main += 4 if sections.get("summary") else 1
    main += 5 if sections.get("skills") else 2
    main += 4 if parsed["word_count"] < 900 else 3
    main += 2.5 if links else 1.5
    main += 1.5 if contact["email_found"] else 0.5

    internal += 5 if len(bullets) >= 5 else 3
    internal += 4
    internal += 3.5 if sections.get("skills") else 2
    internal += 3.5 if sections.get("education") else 1.5
    internal += 4
    internal += 2.5 if sections.get("education") else 1
    internal += 3 if links else 2
    internal += 2

    if not target_role:
        issues.append({
            "severity": "high",
            "title": "Target role is not specified" if lang == "en" else "الوظيفة المستهدفة غير محددة",
            "why": "The analysis is generic and cannot fully judge role fit." if lang == "en" else "التحليل سيكون عامًا ولا يمكنه قياس التوافق الوظيفي بدقة.",
            "action": "Add a target role for a sharper evaluation." if lang == "en" else "أضف وظيفة مستهدفة للحصول على تقييم أدق.",
        })
    if not links:
        issues.append({
            "severity": "critical",
            "title": "No portfolio or project evidence" if lang == "en" else "لا توجد روابط أو أدلة للمشاريع",
            "why": "Claims are harder to verify during first screening." if lang == "en" else "يصعب على مسؤول التوظيف التحقق من أقوى ما في السيرة.",
            "action": "Add a portfolio, GitHub, demo, or case study link." if lang == "en" else "أضف Portfolio أو GitHub أو Demo أو Case Study.",
        })
    if not sections.get("education"):
        issues.append({
            "severity": "high",
            "title": "Education section is missing or unclear" if lang == "en" else "قسم التعليم مفقود أو غير واضح",
            "why": "Education is important for students and early candidates." if lang == "en" else "التعليم مهم جدًا للطلاب والمبتدئين.",
            "action": "Add degree, institution, field, and expected graduation." if lang == "en" else "أضف الدرجة، الجامعة، التخصص، والتخرج المتوقع.",
        })

    if sections.get("skills"):
        strengths.append({"title": "Skills section detected" if lang == "en" else "قسم المهارات موجود", "evidence": "Technical skills / Skills"})
    if bullets:
        strengths.append({"title": "Bullet points found" if lang == "en" else "توجد نقاط خبرة قابلة للتحليل", "evidence": bullets[0][:160]})
    if numbers:
        strengths.append({"title": "Numeric evidence exists" if lang == "en" else "توجد أرقام داخل السيرة", "evidence": ", ".join(numbers[:5])})

    main = min(60, round(main, 1))
    internal = min(40, round(internal, 1))
    final = min(100, max(0, round(main + internal, 1)))

    if not target_role:
        final = min(final, 84)
    if not links:
        final = min(final, 84)

    decision_ar = "يحتاج إلى تعديلات جوهرية قبل التقديم" if final < 75 or any(i["severity"] == "critical" for i in issues) else "قابل للتقديم بعد تعديلات محدودة"
    decision_en = "Needs major revisions before applying" if final < 75 or any(i["severity"] == "critical" for i in issues) else "Submit after limited edits"

    return {
        "scores": {"main_score": main, "internal_score": internal, "penalties": 0, "final_score": final},
        "classification": classify(final, lang),
        "decision": decision_ar if lang == "ar" else decision_en,
        "issues": issues[:5],
        "strengths": strengths[:3],
        "ats": {
            "estimated_score": 76 if contact["email_found"] and sections.get("skills") else 58,
            "readable_text": True,
            "keyword_stuffing": False,
        },
    }

File: frontend/lib/server/evaluator.ts
import { AnalysisMode, Evaluation, Issue, ParsedCv, ReportLanguage, ScoreItem } from "./types";

function clean(value?: string | null) {
  return (value || "").trim();
}

export function getAnalysisMode(targetRole?: string | null, jobDescription?: string | null): AnalysisMode {
  if (clean(jobDescription)) return "targeted";
  if (clean(targetRole)) return "role_only";
  return "generic";
}

function classify(score: number, lang: ReportLanguage) {
  const labels = lang === "ar"
    ? [[95, "استثنائية ونادرة"], [90, "ممتازة جدًا"], [85, "قوية"], [80, "جيدة"], [70, "متوسطة"], [60, "ضعيفة نسبيًا"], [50, "غير مقنعة"], [0, "غير جاهزة مهنيًا"]] as const
    : [[95, "Exceptional"], [90, "Very strong"], [85, "Strong"], [80, "Good"], [70, "Average"], [60, "Relatively weak"], [50, "Unconvincing"], [0, "Not ready"]] as const;
  return labels.find(([threshold]) => score >= threshold)?.[1] || labels[labels.length - 1][1];
}

function issue(lang: ReportLanguage, severity: Issue["severity"], titleEn: string, titleAr: string, whyEn: string, whyAr: string, actionEn: string, actionAr: string): Issue {
  return {
    severity,
    title: lang === "ar" ? titleAr : titleEn,
    why: lang === "ar" ? whyAr : whyEn,
    action: lang === "ar" ? actionAr : actionEn,
  };
}

function scoreItem(criterion: string, maxScore: number, score: number, reason: string, priority: string): ScoreItem {
  return {
    criterion,
    max_score: maxScore,
    score: Math.round(Math.min(Math.max(score, 0), maxScore) * 10) / 10,
    reason,
    priority,
  };
}

function keywordOverlap(text: string, jobDescription?: string | null) {
  const jd = clean(jobDescription).toLowerCase();
  if (!jd) return 0;
  const words = new Set((jd.match(/[a-zA-Z][a-zA-Z+#.]{2,}/g) ?? []).filter((word) => word.length > 3));
  if (!words.size) return 0;
  const cvWords = new Set(text.toLowerCase().match(/[a-zA-Z][a-zA-Z+#.]{2,}/g) ?? []);
  return Array.from(words).filter((word) => cvWords.has(word)).length / words.size;
}

export function evaluate(parsed: ParsedCv, targetRole: string | null, jobDescription: string | null, lang: ReportLanguage): Evaluation {
  const sections = parsed.sections;
  const contact = parsed.contact_information;
  const bullets = parsed.bullet_points;
  const numbers = parsed.numbers;
  const links = contact.links;
  const mode = getAnalysisMode(targetRole, jobDescription);
  const searchableText = [bullets.join(" "), clean(targetRole), clean(jobDescription)].join(" ").toLowerCase();
  const overlap = keywordOverlap(searchableText, jobDescription);

  const issues: Issue[] = [];
  const strengths: Evaluation["strengths"] = [];
  const scoreCaps: Evaluation["score_caps"] = [];
  const hasWorkEvidence = Boolean(sections.experience || sections.projects);
  const hasProjectSection = Boolean(sections.projects);
  const hasNumbers = numbers.length > 0;
  const sectionCount = Object.values(sections).filter(Boolean).length;

  const main = [
    scoreItem("job_alignment", 10, ({ targeted: 8.5, role_only: 7, generic: 4 }[mode]) + Math.min(1.5, overlap * 3), "Mode and target clarity", mode === "generic" ? "high" : "medium"),
    scoreItem("experience_achievements", 10, (hasWorkEvidence ? 4 : 1.5) + Math.min(4, bullets.length * 0.7) + (hasNumbers ? 2 : 0), "Experience/project sections and achievement signals", "high"),
    scoreItem("measurable_results", 7, Math.min(7, numbers.length * 1.2), "Visible numeric evidence only", hasNumbers ? "medium" : "high"),
    scoreItem("titles_dates_credibility", 5, (parsed.dates.length ? 2.5 : 1) + (hasWorkEvidence ? 1.5 : 0.5) + (contact.email_found ? 1 : 0), "Dates, roles, and contact extractability", "medium"),
    scoreItem("professional_summary", 6, sections.summary ? 5 : 1.5, "Professional summary presence", "medium"),
    scoreItem("ats_keywords", 7, (sections.skills ? 2 : 0.5) + Math.min(2, sectionCount * 0.25) + (contact.email_found ? 2 : 0.5) + Math.min(1, overlap * 2), "Standard sections, contacts, and JD keyword overlap", "medium"),
    scoreItem("ten_second_readability", 5, (parsed.word_count <= 900 ? 2 : 1) + (sectionCount >= 4 ? 1.5 : 0.5) + (bullets.length >= 3 ? 1.5 : 0.5), "Skimmability from section and bullet structure", "medium"),
    scoreItem("unsupported_claims", 4, Math.max(0.5, 4 - parsed.buzzwords.length * 0.8), "Conservative penalty for unsupported inflated wording", parsed.buzzwords.length ? "medium" : "low"),
    scoreItem("project_evidence", 3, (hasProjectSection ? 1.5 : 0.5) + (links.length ? 1.5 : 0), "Project section and proof links", hasProjectSection && !links.length ? "critical" : "medium"),
    scoreItem("general_professionalism", 3, (contact.email_found ? 1 : 0.3) + (contact.phone_found ? 0.8 : 0.3) + (parsed.word_count <= 1200 ? 1.2 : 0.6), "Basic professional details", "low"),
  ];

  const internal = [
    scoreItem("bullet_quality", 8, Math.min(8, bullets.length * 0.9 + (hasNumbers ? 2 : 0)), "Bullet count and measurable evidence", "high"),
    scoreItem("language_grammar", 5, Math.max(2, 5 - parsed.buzzwords.length * 0.45), "No grammar AI is used in fallback; inflated wording is penalized", "medium"),
    scoreItem("skill_depth", 5, sections.skills && hasWorkEvidence ? 3.5 : sections.skills ? 2 : 1, "Skills are not treated as proof without evidence", "high"),
    scoreItem("section_structure", 5, Math.min(5, sectionCount * 0.75), "Recognized section structure", "medium"),
    scoreItem("visual_consistency", 5, parsed.line_count >= 8 && parsed.word_count <= 1200 ? 3.8 : 2.5, "Text extraction gives limited formatting evidence", "medium"),
    scoreItem("education_certifications", 4, (sections.education ? 2.5 : 0.8) + (sections.certifications ? 1.5 : 0), "Education and certification sections", "medium"),
    scoreItem("recruiter_risks", 5, 5 - (mode === "generic" ? 1.5 : 0) - (hasProjectSection && !links.length ? 1.5 : 0) - (hasNumbers ? 0 : 1), "Likely first-screening objections", "high"),
    scoreItem("small_details", 3, (contact.email_found ? 1 : 0.2) + (links.length ? 1 : 0.4) + (parsed.dates.length ? 1 : 0.4), "Small details visible in extracted text", "low"),
  ];

  if (mode === "generic") {
    issues.push(issue(lang, "high", "Target role is not specified", "الوظيفة المستهدفة غير محددة", "The analysis is generic and cannot fully judge role fit.", "التحليل عام ولا يمكنه قياس التوافق مع وظيفة محددة بدقة.", "Add a target role for a sharper evaluation.", "أضف وظيفة مستهدفة للحصول على تقييم أدق."));
    scoreCaps.push({ cap: 84, reason: "No target role or job description" });
  } else if (mode === "role_only") {
    scoreCaps.push({ cap: 89, reason: "Target role exists without job description" });
  }

  if (hasProjectSection && !links.length) {
    issues.push(issue(lang, "critical", "Projects have no proof links", "المشاريع بلا روابط إثبات", "Project claims are harder to verify without GitHub, portfolio, demo, or case-study links.", "يصعب التحقق من قوة المشاريع بدون GitHub أو Portfolio أو Demo أو Case Study.", "Add proof links and label what each link proves.", "أضف روابط إثبات ووضح ما يثبته كل رابط."));
    scoreCaps.push({ cap: 84, reason: "Projects are core evidence but no proof link exists" });
  }
  if (!hasNumbers) {
    issues.push(issue(lang, "high", "No measurable results", "لا توجد نتائج قابلة للقياس", "The CV describes content without showing measurable impact or scale.", "السيرة تعرض محتوى بدون أثر أو حجم عمل قابل للقياس.", "Add real numbers only where the CV can support them.", "أضف أرقامًا حقيقية فقط عندما يمكن إثباتها."));
    scoreCaps.push({ cap: 84, reason: "No measurable impact metrics" });
  }
  if (!sections.education) {
    issues.push(issue(lang, "high", "Education section is missing or unclear", "قسم التعليم مفقود أو غير واضح", "Education is important for students, fresh graduates, and entry-level applicants.", "التعليم مهم للطلاب والخريجين والمبتدئين.", "Add degree, institution, field, and graduation date/status.", "أضف الدرجة، المؤسسة، التخصص، وتاريخ أو حالة التخرج."));
  }

  if (sections.skills) strengths.push({ title: lang === "ar" ? "قسم المهارات موجود" : "Skills section detected", evidence: "Skills / Technical Skills" });
  if (bullets.length) strengths.push({ title: lang === "ar" ? "توجد نقاط قابلة للتحليل" : "Analyzable bullet points found", evidence: bullets[0].slice(0, 160) });
  if (numbers.length) strengths.push({ title: lang === "ar" ? "توجد أرقام داخل السيرة" : "Numeric evidence exists", evidence: numbers.slice(0, 5).join(", ") });
  if (links.length) strengths.push({ title: lang === "ar" ? "يوجد رابط إثبات" : "Proof link detected", evidence: links[0].slice(0, 120) });

  const mainScore = Math.min(60, Math.round(main.reduce((total, item) => total + item.score, 0) * 10) / 10);
  const internalScore = Math.min(40, Math.round(internal.reduce((total, item) => total + item.score, 0) * 10) / 10);
  const baseScore = Math.round((mainScore + internalScore) * 10) / 10;
  const capValue = Math.min(100, ...scoreCaps.map((cap) => cap.cap));
  const finalScore = Math.min(100, Math.max(0, Math.round(Math.min(baseScore, capValue) * 10) / 10));
  const penalties = Math.round((baseScore - finalScore) * 10) / 10;

  const decisionKey = finalScore >= 85 && hasNumbers && links.length && mode !== "generic" ? "ready" : finalScore >= 75 && !issues.some((item) => item.severity === "critical") ? "limited" : finalScore >= 50 ? "major" : "rebuild";
  const decisions = {
    ar: { ready: "جاهز للتقديم الآن", limited: "قابل للتقديم بعد تعديلات محدودة", major: "يحتاج إلى تعديلات جوهرية قبل التقديم", rebuild: "يحتاج إلى إعادة بناء كاملة" },
    en: { ready: "Ready to apply now", limited: "Submit after limited edits", major: "Needs major revisions before applying", rebuild: "Needs a full rebuild" },
  };

  return {
    analysis_mode: mode,
    scores: { main_score: mainScore, internal_score: internalScore, penalties, final_score: finalScore },
    score_caps: scoreCaps,
    criteria: { main, internal },
    classification: classify(finalScore, lang),
    decision: decisions[lang][decisionKey],
    issues: issues.slice(0, 5),
    strengths: strengths.slice(0, 3),
    ats: {
      estimated_score: Math.min(100, Math.round((main[5].score / 7) * 100)),
      readable_text: true,
      keyword_stuffing: parsed.buzzwords.length >= 4,
      standard_sections_found: Object.entries(sections).filter(([, present]) => present).map(([key]) => key),
      jd_keyword_overlap: jobDescription ? Math.round(overlap * 100) / 100 : null,
    },
  };
}
